programs/castle-vault/src/asset_container/mod.rs

```diff
@@ -0,0 +1,123 @@
+mod iter;
+mod rate;
+mod reserves;
+mod u64;
+
+pub use self::u64::*;
+pub use iter::*;
+pub use rate::*;
+pub use reserves::*;
+
+use core::ops::{Index, IndexMut};
+
+use strum::{EnumCount, IntoEnumIterator};
+
+use crate::reserves::Provider;
+
+pub type AssetContainer<T> = AssetContainerGeneric<T, { Provider::COUNT }>;
+
+/// Provides an abstraction over supported assets
+#[derive(Debug, Clone)]
+pub struct AssetContainerGeneric<T, const N: usize> {
+    pub(crate) inner: [Option<T>; N],
+}
+
+impl<T, const N: usize> AssetContainerGeneric<T, N> {
+    pub fn len(&self) -> usize {
+        N
+    }
+
+    pub fn is_empty(&self) -> bool {
+        // TODO: Should this just always return `true`, or do we want it to mean "uninitialized"?
```

Comment:
May be worth adding a comment communicating that `is_empty()` means "uninitialized"